ext/random/random.c

```diff
@@ -680,35 +680,82 @@ uint64_t php_random_generate_fallback_seed(void)
 		gettimeofday(&tv, NULL);
 		fallback_seed_add(&c, &tv, sizeof(tv));
 		/* Previous state. */
-		fallback_seed_add(&c, RANDOM_G(fallback_seed), 20);
+		fallback_seed_add(&c, state->seed, 20);
 	}
-	PHP_SHA1Final(RANDOM_G(fallback_seed), &c);
-	RANDOM_G(fallback_seed_initialized) = true;
+	PHP_SHA1Final(state->seed, &c);
+	state->initialized = true;
 
 	uint64_t result = 0;
 
 	for (int i = 0; i < sizeof(result); i++) {
-		result = result | (((uint64_t)RANDOM_G(fallback_seed)[i]) << (i * 8));
+		result = result | (((uint64_t)state->seed[i]) << (i * 8));
 	}
 
 	return result;
 }
 
-/* {{{ PHP_GINIT_FUNCTION */
-static PHP_GINIT_FUNCTION(random)
+uint64_t php_random_generate_fallback_seed(void)
 {
-	random_globals->random_fd = -1;
-	random_globals->fallback_seed_initialized = false;
+	return php_random_generate_fallback_seed_ex(&RANDOM_G(fallback_seed_state));
 }
-/* }}} */
 
-/* {{{ PHP_GSHUTDOWN_FUNCTION */
-static PHP_GSHUTDOWN_FUNCTION(random)
+static zend_result php_general_random_bytes_for_zend_initialize(php_random_state_for_zend *state)
+{
+	uint64_t t[4];
+
+	do {
+		char errstr[128];
+		if (php_random_bytes_ex(&t, sizeof(t), errstr, sizeof(errstr)) == FAILURE) {
+#if ZEND_DEBUG
+			fprintf(stderr, "php_random_bytes_ex: Failed to generate a random seed: %s\n", errstr);
+#endif
+			return FAILURE;
+		}
+	} while (UNEXPECTED(t[0] == 0 && t[1] == 0 && t[2] == 0 && t[3] == 0));
+
+	php_random_xoshiro256starstar_seed256(&state->xoshiro256starstar_state, t[0], t[1], t[2], t[3]);
+
+	return SUCCESS;
+}
+
+static void php_general_random_bytes_for_zend_initialize_fallback(php_random_state_for_zend *state)
+{
+	uint64_t t;
+	php_random_fallback_seed_state fallback_state;
+
+	do {
+		t = php_random_generate_fallback_seed_ex(&fallback_state);
+	} while (UNEXPECTED(t == 0));
+
+	php_random_xoshiro256starstar_seed64(&state->xoshiro256starstar_state, t);
+}
+
+PHPAPI zend_result php_general_random_bytes_for_zend(zend_utility_general_random_state *opaque_state, void *bytes, size_t size)
```

Comment:
Not a fan of the method name: `random_bytes()` is understood as guaranteeing the CSPRNG for both userland and internals. This one uses insecure randomness and might lead to accidental misuse, especially given it's part of the regular `php_random.h`.

Consider either renaming the function, or moving it into a dedicated header file (or both).

Reply:
Do you have any suggestion for a name? I've used `general` as in "general purpose". Maybe php_weak_random_bytes or php_insecure_random_bytes?

Reply:
Moving the `_zend` functions into a new `php_random_zendhelpers.h` or so would probably make sense nevertheless. The functions are probably not useful to non-Zend users and having them in a separate headers keeps `php_random.h` nice and tidy.

Reply:
Thanks. I moved these to `ext/random/zend_utils.c` / `php_random_zend_utils.h`, and renamed the functions as suggested. I've added you as author, since your wrote most of this code, but let me know if you don't want that.

I also renamed the `zend_os_csprng_random_bytes` utility to just `zend_random_bytes`.

Reply:
> I've added you as author, since your wrote most of this code, but let me know if you don't want that.

That works for me, thanks.